### ISSUE TITLE
Tracking issue for customizing CRT linkage

### ISSUE THREAD

**Issue by alexcrichton** [rust-lang/rust#37406]
Tracking issue for https://github.com/rust-lang/rfcs/pull/1721
- [x] - Add the `crt-static` target-feature
- [x] - Update Cargo to pass through target-feature to build scripts - https://github.com/rust-lang/cargo/pull/3243
- [x] - Implement "lazy link attributes" and implement in a gated fashion in liblibc.

**Comment by retep998** [comment]
I can confirm that with an up to date libc, using this new feature out in the wild does result in binaries with the CRT statically linked.

![](http://i.imgur.com/iWLf0Yw.png)

**Comment by steveklabnik** [comment]
So this landed in stable right? Can this be closed now?

**Comment by alexcrichton** [comment]
AFAIK this feature is still gated, but I'll update the tags as such.

**Comment by tombruijn** [comment]
Compiling a staticlib with musl using this feature (`RUSTFLAGS='-C target-feature=-crt-static -Z unstable-options' cargo build --target=x86_64-unknown-linux-musl`) allows it to be properly linked when installing it with a Ruby gem. A problem I reported in #37007. A musl compiled binary already worked without this feature, but turning it on will break it.

We're currently running a tests on an alpha release of the AppSignal gem using a staticlib compiled this way (https://github.com/appsignal/appsignal-ruby/pull/229) and so far we have not experienced any problems with the musl compiled staticlib.

Thanks for the hard work!

**Comment by alexcrichton** [comment]
@rfcbot fcp merge

I'd like to move this issue to FCP for stabilization as we've got some good traction using this and it looks to be working quite well.

**Comment by alexcrichton** [comment]
ping @japaric (checkbox)

**Comment by japaric** [comment]
@alexcrichton Thanks for the ping and sorry for holding this up 😅.

**Comment by alexcrichton** [comment]
No worries, thanks @japaric!

**Comment by alexcrichton** [comment]
Discussed with @rust-lang/tools the other day and the conclusion was that we should switch the compiler itself over to using libcmt instead of msvcrt, and then we should be good to stabilize.

**Comment by retep998** [comment]
The only question is whether we're passing any CRT objects around, because rustc is dynamically linked. libstd doesn't rely on CRT state, but LLVM might. Any code that works with CRT objects *has* to stay within a given dll, it cannot be inlined or monomorphized into other dlls or bad things will happen.

**Comment by alexcrichton** [comment]
I've made a PR for changing how the compiler is compiled: https://github.com/rust-lang/rust/pull/39837

**Comment by aturon** [comment]
cc @rust-lang/compiler -- anyone want to make a quick stabilization PR that we can backport to 1.17 beta?

**Comment by nikomatsakis** [comment]
Stabilization is a great place for a first PR. There are directions available on the forge:

https://forge.rust-lang.org/stabilization-guide.html

However, @alexcrichton -- are there docs available for this feature?

**Comment by steveklabnik** [comment]
I'm not aware of any.

**Comment by alexcrichton** [comment]
I've added documentation in https://github.com/rust-lang-nursery/reference/pull/17

In doing so I've realized that we have [another can of worms to open](https://github.com/rust-lang/rust/issues/29717) before we can stabilize this, unfortunately.

**Comment by binarycrusader** [comment]
Be aware that some platforms do not offer a static libc, such as Solaris. Fyi.

**Comment by alexcrichton** [comment]
This was discussed at tools triage recently and the decision was to go ahead and stabilize `-C target-feature=+crt-static`, but we will avoid stabilizing `#[cfg]` accessors as that's connected [elsewhere](https://github.com/rust-lang/rust/issues/29717)

**Comment by alexcrichton** [comment]
I've opened a [stabilization PR](https://github.com/rust-lang/rust/pull/41757) to reflect this conclusion.

**Comment by clarcharr** [comment]
So apparently, libc fails to compile on MUSL for me because of this feature:

```
error: is feature gated (see issue #37406)
   --> /home/clar/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.26/src/unix/mod.rs:227:9
    |
227 |         #[link(name = "c", kind = "static", cfg(target_feature = "crt-static"))]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(link_cfg)] to the crate attributes to enable

error: is feature gated (see issue #37406)
   --> /home/clar/.cargo/registry/src/github.com-1ecc6299db9ec823/libc-0.2.26/src/unix/mod.rs:228:9
    |
228 |         #[link(name = "c", cfg(not(target_feature = "crt-static")))]
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = help: add #![feature(link_cfg)] to the crate attributes to enable
```

But this appears to be stabilised? Am I missing something?

**Comment by steveklabnik** [comment]
@clarcharr I believe that's because it's stable in 1.19, which is gonna get released in like 45 minutes.

**Comment by clarcharr** [comment]
@steveklabnik sorry, I should have clarified that this is on nightly.

**Comment by QuietMisdreavus** [comment]
It looks like the issue is more about using the `cfg(target_feature)`, rather than `crt-static` itself? Since #29717 is still open, i bet that's going on.

**Comment by retep998** [comment]
`libc` should not be relying on the `link_cfg` feature except when built as part of the standard library. When built as a normal crates.io crate that attribute should not exist, which means it is likely a bug in `libc`.

**Comment by clarcharr** [comment]
@retep998 I'm having trouble imagining this as a bug in libc considering how musl is a tested target; maybe a recent change triggered this?

**Comment by retep998** [comment]
The syntax `#[link(name = "c", kind = "static", cfg(target_feature = "crt-static"))]` is intended to be an unstable detail only used for `libc` when built as part of the standard library because the standard library is compiled at a point where it does not know whether the user wants a static CRT or dynamic CRT.

As far as I can tell by looking at the code, that attribute should only even be activated during a cargo build when you've opted out of the `use_std` default feature. However I see that CI does test `cargo build --no-default-features` which makes me wonder whether it is testing that target. Because it doesn't explicitly specify `--target` that means.... it isn't actually testing that target until it does the docker tests, which never build the crate with `--no-default-features`!

**Comment by clarcharr** [comment]
aha!

I'll link this in a comment on libc to point out that the musl build is indeed failing.

**Comment by varkor** [comment]
This issue is still linked to for `#[link(cfg(foo))]`, so shouldn't it still be open?

**Comment by petrochenkov** [comment]
@varkor 
>This issue is still linked to in errors for `#[link(cfg(foo))]`, so shouldn't it still be open?

It's better to make a new issue (or remove reference to this issue from the feature gate).

**Comment by varkor** [comment]
I've opened https://github.com/rust-lang/rust/issues/72059, though I haven't had time to look into the state of this feature.
